File: Array/Binary Search/35_Search_Insert_Position.cpp
#include<iostream>
#include<string>
#include<vector>

using namespace std;

int searchInsert(vector<int>& nums, int target) {
    int res = -1;

    int left = 0, right = nums.size()-1;
    int mid = left+ (right - left)/2;

    while(true){
        if(left<right){
            if(nums[mid] == target){
                res = mid;
                break;
            }else if(nums[mid]<target){
                left = mid + 1;
            }else{
                right = mid;
            }
        }else if(left >= right){
            if(nums[mid] == target){
                res = mid;
                break;
            }else if(nums[mid]<target){
                res = mid +1;
                break;
            }else{
                res = mid;
                break;
            }
        }

        mid = left+ (right - left)/2;
    }


    return res;
}


int main(){
    vector<int> test = {-3, 0, 1, 3,9, 10};
    cout<<searchInsert(test, 2)<<endl;
    return 0;
}

File: Array/Binary Search/35_Search_Insert_Position.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
var searchInsert = function (nums, target) {
  let l = 0,
    r = nums.length - 1;

  while (r >= l) {
    const mid = Math.floor((r + l) / 2);
    const midN = nums[mid];

    if (midN === target) {
      return mid;
    } else if (target > midN) {
      l = mid + 1;
    } else {
      r = mid - 1;
    }
  }

  return l - 1;
};

console.log(searchInsert([1, 3, 5, 6], 4));
